gapil/compiler/plugins/replay/replay.go

```diff
@@ -85,314 +292,118 @@ func (r *replayer) emitCall(s *compiler.S) {
-			default:
-				r.Fail("Unsupported integer size: %v", size)
-			}
-		case semantic.SizeType:
-			size := r.replayABI.MemoryLayout.Size.Size
-			switch size {
-			case 8:
-				r.writePush(s, val, protocol.Type_Uint8)
-			case 16:
-				r.writePush(s, val, protocol.Type_Uint16)
-			case 32:
-				r.writePush(s, val, protocol.Type_Uint32)
-			case 64:
-				r.writePush(s, val, protocol.Type_Uint64)
-			default:
-				r.Fail("Unsupported size_t size: %v", size)
-			}
-		case semantic.CharType:
-			size := r.replayABI.MemoryLayout.Char.Size
-			switch size {
-			case 8:
-				r.writePush(s, val, protocol.Type_Uint8)
-			case 16:
-				r.writePush(s, val, protocol.Type_Uint16)
-			case 32:
-				r.writePush(s, val, protocol.Type_Uint32)
-			case 64:
-				r.writePush(s, val, protocol.Type_Uint64)
-			default:
-				r.Fail("Unsupported char_t size: %v", size)
-			}
-		case semantic.Int8Type:
-			r.writePush(s, val, protocol.Type_Int8)
-		case semantic.Uint8Type:
-			r.writePush(s, val, protocol.Type_Uint8)
-		case semantic.Int16Type:
-			r.writePush(s, val, protocol.Type_Int16)
-		case semantic.Uint16Type:
-			r.writePush(s, val, protocol.Type_Uint16)
-		case semantic.Int32Type:
-			r.writePush(s, val, protocol.Type_Int32)
-		case semantic.Uint32Type:
-			r.writePush(s, val, protocol.Type_Uint32)
-		case semantic.Int64Type:
-			r.writePush(s, val, protocol.Type_Int64)
-		case semantic.Uint64Type:
-			r.writePush(s, val, protocol.Type_Uint64)
+			return r.asmBool(s, val)
+
+		case semantic.IntType, semantic.UintType, semantic.SizeType, semantic.CharType,
+			semantic.Int8Type, semantic.Uint8Type,
+			semantic.Int16Type, semantic.Uint16Type,
+			semantic.Int32Type, semantic.Uint32Type,
+			semantic.Int64Type, semantic.Uint64Type:
+			return r.asmValue(s, val.Cast(r.T.Uint64), r.asmType(ty))
+
 		case semantic.Float32Type:
-			r.writePush(s, val, protocol.Type_Float)
+			return r.asmValue(s, val.Bitcast(r.T.Uint32).Cast(r.T.Uint64), r.asmType(ty))
+
 		case semantic.Float64Type:
-			r.writePush(s, val, protocol.Type_Double)
+			return r.asmValue(s, val.Bitcast(r.T.Uint64), r.asmType(ty))
 		}
-	case *semantic.StaticArray:
-		r.Fail("TODO")
 	case *semantic.Pointer:
-		r.Fail("TODO")
-	case *semantic.Class:
-		r.Fail("TODO")
+		return s.Select(s.LessThan(val, s.Scalar(observableAddressStart).Cast(val.Type())),
+			r.asmAbsolutePtr(s, val),
+			r.observedPtr(s, val, ty))
 	}
+	r.Fail("Unhandled type %v", ty.Name())
+	return nil
 }
 
-const (
-	// Various bit-masks used by opcode writing.
-	// Many opcodes can fit values into the opcode itself.
-	// These masks are used to determine which values fit.
-
-	mask19 = uint64(0x7ffff)
-	mask20 = uint64(0xfffff)
-	mask26 = uint64(0x3ffffff)
-	mask45 = uint64(0x1fffffffffff)
-	mask46 = uint64(0x3fffffffffff)
-	mask52 = uint64(0xfffffffffffff)
-
-	//     ▏60       ▏50       ▏40       ▏30       ▏20       ▏10
-	// ○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○○●●●●●●●●●●●●●●●●●●● mask19
```

Comment:
I'm going to miss all of the unicode drawings for replay.

Reply:
Don't worry. They've just been moved to `gapil/runtime/cc/replay/builder.cpp`. 😃